tests/Integration/Command/Share/TShareCommandTest.php

```diff
@@ -0,0 +1,129 @@
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Polls\Tests\Integration\Command\Share;
+
+use OCA\Polls\Db\Poll;
+use OCA\Polls\Db\PollMapper;
+use OCA\Polls\Db\Share;
+use OCA\Polls\Db\ShareMapper;
+use OCA\Polls\Service\ShareService;
+use OCP\IGroupManager;
+use OCP\IUserManager;
+use PHPUnit\Framework\MockObject\MockObject;
+
+trait TShareCommandTest {
+	/** @var PollMapper|MockObject */
+	protected $pollMapper;
+
+	/** @var ShareMapper|MockObject */
+	protected $shareMapper;
+
+	/** @var ShareService|MockObject */
+	protected $shareService;
+
+	/** @var IUserManager|MockObject */
+	protected $userManager;
+
+	/** @var IGroupManager|MockObject */
+	protected $groupManager;
+
+	/** @var int */
+	protected $lastShareId = 0;
+
+	protected function setUpMocks(): void {
+		$this->pollMapper = $this->createMock(PollMapper::class);
+		$this->shareMapper = $this->createMock(ShareMapper::class);
+		$this->shareService = $this->createMock(ShareService::class);
+		$this->userManager = $this->createMock(IUserManager::class);
+		$this->groupManager = $this->createMock(IGroupManager::class);
+	}
+
+	public function createPollMock(int $id): Poll {
+		/** @var Poll|MockObject $poll */
+		$poll = $this->createMock(
+			Poll::class,
+			['getId']
+		);
+
+		$poll->method('getId')
+			->willReturn($id);
+
+		return $poll;
+	}
+
+	public function createShareMock(int $pollId, string $type, string $userId): Share {
+		/** @var Share|MockObject $share */
+		$share = $this->createMock(
+			Share::class,
+			['getId', 'getPollId', 'getType', 'getEmailAddress', 'getToken'],
+			['getUserId']
+		);
+
+		$id = ++$this->lastShareId;
+		$token = $this->getShareToken($pollId, $type, $userId);
+
+		$share->method('getId')
+			->willReturn($id);
+
+		$share->method('getPollId')
+			->willReturn($pollId);
+
+		$share->method('getType')
+			->willReturn($type);
+
+		$share->method('getUserId')
+			->willReturn($userId);
+
+		$share->method('getEmailAddress')
+			->willReturn($userId);
+
+		$share->method('getToken')
+			->willReturn($token);
+
+		return $share;
+	}
+
+	public function getShareToken(int $pollId, string $type, string $userId): string {
+		return substr(md5($pollId . '_' . $type . '_' . $userId), 0, 16);
+	}
+
+	protected function createMock($originalClassName, array $addMethods = null, array $onlyMethods = null): MockObject {
+		$mockBuilder = $this->getMockBuilder($originalClassName)
+			->disableOriginalConstructor()
+			->disableOriginalClone()
+			->disableArgumentCloning()
+			->disallowMockingUnknownTypes()
+			->disableAutoReturnValueGeneration();
+
+		if ($addMethods !== null) {
+			$mockBuilder->addMethods($addMethods);
+		}
+
+		if ($onlyMethods !== null) {
+			$mockBuilder->onlyMethods($onlyMethods);
+		}
```

Comment:
By the way: I'd highly recommend not to use magic methods for getters/setters. Explicitly adding the methods to mock is only necessary because of magic methods.

Reply:
I have no experience with this. The code for the tests isn't created by me. Do you mind to optimize that?

Reply:
This is not about the test's code :smile:

The issue resides within your model classes. Just take the `OCA\Polls\Db\Share::getId()` method as an example: There's no `OCA\Polls\Db\Share::getId()` method, there's only a `OCP\AppFramework\Db\Entity::__call()` method that "magically" returns the requested property. Using magic methods is discouraged, because they are incredibly hard to test. This code snippet (also see [[1]](https://github.com/nextcloud/polls/pull/1516/files#diff-6a85ee1ca2ddc000e18ce3f0c0a8d2104197e04d4e0badb9a5c9effbff294462R66) and [[2]](https://github.com/nextcloud/polls/pull/1516/files#diff-6a85ee1ca2ddc000e18ce3f0c0a8d2104197e04d4e0badb9a5c9effbff294462R79-R80)) works around this issue by manually specifying all methods to mock - not just the magic methods, but *all* methods. This is bad. It works, but is nevertheless bad. It's the only solution because you're using magic methods. You should rather implement the getters you need (i.e. e.g. `OCA\Polls\Db\Share::getId()`).

Anyway, this is just a heads up for your. Fixing this technical debt is not in the scope of this PR. I can open a separate issue if you want to.

Reply:
Ah. I thought it was about this implementation: 
https://github.com/nextcloud/polls/blob/09b4c9006b070ae7e36f9f4cf947a02069980d45/tests/Unit/FactoryMuffin.php#L19-L32

The style of magic getters and setters follows the NC style guide IMHO. Wasn't questioned by me. 

https://docs.nextcloud.com/server/latest/developer_manual/basics/storage/database.html#entities

THB, this project was my first bigger developing experience.

Reply:
Another workaround for magic getters/setters :smile: This is nothing that needs to be addressed *right now*, you just might want to keep it in mind.